APLWKWebView/APLWKWebViewController.m

```diff
@@ -344,23 +268,120 @@ - (UIBarButtonItem *)forwardButtonItem {
+                                                                          target:self.webView
+                                                                          action:@selector(reload)];
+    }
+    return _reloadButtonItem;
+}
 
-- (NSMutableArray *)pendingLoadThresholdReachedCompletionHandlers {
-    if (!_pendingLoadThresholdReachedCompletionHandlers) {
-        _pendingLoadThresholdReachedCompletionHandlers = [NSMutableArray new];
+#pragma mark - KVO: Loading Progress
+
+- (void)observeWebView:(WKWebView *)webView {
+    [webView addObserver:self forKeyPath:@"estimatedProgress" options:NSKeyValueObservingOptionNew context:kAPLWKWebViewKVOContext];
+    [webView addObserver:self forKeyPath:@"title" options:NSKeyValueObservingOptionNew context:kAPLWKWebViewKVOContext];
+    [webView addObserver:self forKeyPath:@"loading" options:NSKeyValueObservingOptionNew context:kAPLWKWebViewKVOContext];
+    [webView addObserver:self forKeyPath:@"canGoBack" options:NSKeyValueObservingOptionNew context:kAPLWKWebViewKVOContext];
+    [webView addObserver:self forKeyPath:@"canGoForward" options:NSKeyValueObservingOptionNew context:kAPLWKWebViewKVOContext];
+}
+
+- (void)removeObserversFromWebView:(WKWebView *)webView {
+    [webView removeObserver:self forKeyPath:@"estimatedProgress" context:kAPLWKWebViewKVOContext];
+    [webView removeObserver:self forKeyPath:@"title" context:kAPLWKWebViewKVOContext];
+    [webView removeObserver:self forKeyPath:@"loading" context:kAPLWKWebViewKVOContext];
+    [webView removeObserver:self forKeyPath:@"canGoBack" context:kAPLWKWebViewKVOContext];
+    [webView removeObserver:self forKeyPath:@"canGoForward" context:kAPLWKWebViewKVOContext];
+
+    if (webView.scrollView.delegate == self) {
+        webView.scrollView.delegate = nil;
     }
+}
 
-    return _pendingLoadThresholdReachedCompletionHandlers;
+- (void)checkDOMReady {
+    if (_didFinishDOMLoad) {
+        return;
+    }
+
+    [self.webView evaluateJavaScript:@"document.readyState != \"loading\"" completionHandler:^(id _Nullable finished, NSError * _Nullable error) {
+        if ([finished boolValue]) {
+            [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
+            [self setProgressViewProgressTo:1 andHideAfter:1.0];
+
+            if ([self->_aplWebViewDelegate respondsToSelector:@selector(aplWebViewController:didChangeLoadingState:)]) {
+                [self->_aplWebViewDelegate aplWebViewController:self didChangeLoadingState:NO];
+            }
+            [self updateBottomItems];
+        }
+    }];
+}
+
+-(void)setProgressViewProgressTo:(CGFloat)progressValue andHideAfter:(NSTimeInterval)delayBeforeHiding {
+    [_progressView setProgress:progressValue animated:YES];
+
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delayBeforeHiding * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+        /*
+         * Imitate Safari, which fills the progress bar before hiding it.
```

Comment:
Wouldn't that require to set progress = 1 somewhere? At least I can see the progress bar to hide sometimes before being filled.

Reply:
@nischu  In the very first line of this method the progress value will be set to the desired value before setting it to 0 and hide the bar entirely after a delay.